Add a helper function for building custom call lowering rules

dfm opened the pull request:
This function provides sensible defaults for custom call lowering rules with the goal of reducing the amount of boilerplate required for implementing custom calls. While all of the behavior can be overridden, this will be particularly useful when (a) the inputs and outputs are in row-major order because then the layouts and types can be evaluated from the context avals, and/or (b) the custom call supports `api_version=4` attributes provided via a `backend_config` dictionary.

There are a number of lowering rules in the core library that could be rewritten using this helper, but I think it probably wouldn't make sense to port those as part of this PR.

For now the tests only check that lowering succeeds for parameters/attributes of different types, but it might be useful to check that the output is appropriate. I haven't figured out a reasonable way to do that yet - thoughts would be appreciated!

@superbobry — I'd love to hear your feedback. Since this is closely based on code you wrote previously, I added you as a co-author on the commit, but I'm happy to remove that if you'd prefer.

dfm committed:
Add helper function for building custom call lowering rules

This function provides sensible defaults for custom call lowering rules
with the goal of reducing the amount of boilerplate required for
implementing custom calls.

Co-authored-by: Sergei Lebedev <slebedev@google.com>

superbobry commented on jax/__init__.py:
```diff
@@ -149,6 +149,7 @@
 
 # These submodules are separate because they are in an import cycle with
 # jax and rely on the names imported above.
+from jax import custom_call as custom_call
```
Should we, perhaps, expose this somewhere under `jax.extend` or even not expose at all until we are sure this is useful? 

CC @froystig

dfm replied:
I think either of these options would make sense. I don't have strong opinions either way!

dfm replied:
I've moved this interface to `jax.extend.custom_call`.

superbobry commented on jax/_src/custom_call.py:
```diff
@@ -0,0 +1,117 @@
+from jax._src import core
+from jax._src import dtypes
+from jax._src.interpreters import mlir
+from jax._src.interpreters.mlir import ir
```
Not sure what the preferred style is, but JFYI I have been importing `ir` from `jax._src.lib.mlir` :)

superbobry commented on jax/_src/custom_call.py:
```diff
@@ -0,0 +1,117 @@
+
+
+def custom_call_lowering(
+  call_target_name: str,
```
Nit: JAX does not strictly follow it, but Google Python style guide recommends 4 spaces for hanging indentation.

superbobry commented on jax/_src/custom_call.py:
```diff
@@ -0,0 +1,117 @@
+  call_target_name: str,
+  api_version: int = 4,
+  backend_config: str | bytes | dict[str, ir.Attribute] | None = None,
+  result_types: Sequence[ir.Type] | None = None,
```
We have an option of using `jax.ShapeDtypeStruct` here instead of MLIR types. Wdyt?

dfm replied:
I haven't used `ShapeDtypeStruct` before so I'm not sure I totally understand what this would mean here? What are the benefits?

superbobry replied:
The benefit is that the caller doesn't reach into the `ir` module to construct MLIR types and can use JAX's dtypes instead.

So, e.g. instead of 

```
result_types=[ir.RankedTensorType([42], ir.IntegerType.get(...)]
result_shapes=[42]
```

the caller can do 

```
result_shapes_dtypes=[jax.ShapeDtypeStruct([42], jnp.float32)]
```

dfm replied:
Got it! In thinking through this, I'm actually inclined to just remove `result_types` and `result_shapes` and not add `result_shapes_dtypes` either. If a user wants override either of these, this helper function really doesn't add much benefit over just calling `mlir.custom_call` directly.

It is still possible to customize these values using `**lowering_args`, but I don't think that they need to be exposed as top level parameters!

superbobry commented on jax/_src/custom_call.py:
```diff
@@ -0,0 +1,117 @@
+
+def custom_call_lowering(
+  call_target_name: str,
+  api_version: int = 4,
```
I wonder if we need to expose `api_version=` here at all. It looks like the new typed FFI is objectively better and more user friendly than its predecessors, so maybe we should just hardcode `api_version=4` and require `backend_config` to be a `dict`/`Mappping`?

superbobry replied:
Otherwise, it is easy for the caller to get unsupported combination of arguments, e.g. 

```
custom_call_lowering(api_version=4, backend_config="foo")
```

dfm replied:
I think agree here with the caveat that maybe we want to support specifying `api_version >= 4` for future compatibility? When the next API version comes out, I'm not sure we'd want a hard coded version here... but I'm not convinced that I know that best way to handle this!

superbobry replied:
Yeah, I don't have a strong opinion here, but maybe @hawkinsp does?

dfm replied:
I think I'm pretty happy just setting the default for `api_version` to `4`. Given the API contract for `extend`, I don't think it's the end of the world to cross this bridge if we get there.

superbobry commented on jax/_src/custom_call.py:
```diff
@@ -0,0 +1,117 @@
+
+def _ir_attribute(obj: Any) -> ir.Attribute:
+  # TODO(dfm): Similar functions exist in Pallas and Mosaic GPU. Perhaps these
+  # could be consolidated into mlir or similar.
```
Definitely!

dfm replied:
I'm inclined to keep this here for now to keep this PR compact, then we could move this function later if it would be useful to consolidate.

superbobry commented on jax/_src/custom_call.py:
```diff
@@ -0,0 +1,117 @@
+    if not (dtypes.is_python_scalar(obj) or np.isscalar(obj)):
+      raise TypeError("Only scalar attributes are supported")
+    mlir_type = mlir.dtype_to_ir_type(dtype)
+    if ir.IntegerType.isinstance(mlir_type) or ir.IndexType.isinstance(mlir_type):
```
FYI you can do `isinstance(..., ir.IntegerType)` here and below for `ir.FloatType`. This is not true for any MLIR types, but it ints/floats have a Python type hierarchy with `ir.IntegerType` and `ir.FloatType` as base classes.

superbobry replied:
Also, I don't think `mlir.dtype_to_ir_type` can return `ir.IndexType`?

superbobry commented on jax/_src/custom_call.py:
```diff
@@ -0,0 +1,117 @@
+  elif isinstance(obj, bool):
+    return ir.BoolAttr.get(obj)
+  else:
+    dtype = dtypes.canonicalize_dtype(np.result_type(obj))
```
I think this will map an `int` either to `i32` or to `i64`, depending on the value of the `enable_x64` flag. This is not what e.g. `mlir.ir_constant` does: https://github.com/google/jax/blob/beaabae8f6d2ceb049840f59c6683e93a67726ae/jax/_src/interpreters/mlir.py#L330.

Should we also unconditionally used `i64` here?

dfm replied:
I've updated the implementation to be consistent with `ir_constant` restricted to scalars.

superbobry commented on jax/_src/extend/ffi.py:
```diff
@@ -15,11 +15,103 @@
   return os.path.join(jaxlib_dir, "include")
+
+
+def ffi_lowering(
```
I like the new name!

superbobry commented on jax/_src/extend/ffi.py:
```diff
@@ -15,11 +15,103 @@
+    backend_config: Mapping[str, ir.Attribute] | None = None,
+    **lowering_args: Any
+) -> mlir.LoweringRule:
+  """Build a lowering rule for an foreign function interface (FFI) target
```
Nit: missing `.`.

superbobry replied:
Should we be mentioning primitives here, or would we point people to the e2e tutorial once we have it?

superbobry commented on jax/_src/extend/ffi.py:
```diff
@@ -15,11 +15,103 @@
+  compute the input and output types and shapes for the custom call, assuming
+  row-major layouts.
+
+  If keyword arguments are passed at bind time, these are treated as attributes,
```
I would guess that "bind time" might not be clear to users. One way to avoid it is to talk about a primitive and its parameters.

We defined primitive [https://jax.readthedocs.io/en/latest/glossary.html#term-primitive] here, but unfortunately don't mention parameters atm.

superbobry commented on jax/_src/extend/ffi.py:
```diff
@@ -15,11 +15,103 @@
+    ctx: mlir.LoweringRuleContext, *operands: ir.Value, **params: Any
+  ) -> Sequence[ir.Value | Sequence[ir.Value]]:
+    kwargs = dict(lowering_args)
+    kwargs["api_version"] = kwargs.get("api_version", 4)
```
Nit: `kwargs.setdefault("api_version", 4)`?

superbobry commented on jax/_src/extend/ffi.py:
```diff
@@ -15,11 +15,103 @@
+    kwargs["backend_config"] = dict(
+      backend_config or {}, **{k: _ir_attribute(v) for k, v in params.items()})
+
+    # TODO(dfm): This is a common pattern for supporting dynamic shapes, but I
```
Not sure if this is really necessary here. CC @gnecula.

dfm replied:
I've just removed this. We can revisit in the future if it's a feature that many users would find useful.

dfm commented:
Thanks for the reviews, @superbobry! I think this is getting close to ready. Here are the main updates:

1. After conversations with @superbobry and @froystig, I agreed that it makes sense for this to live somewhere under `jax.extend`. In particular, I think it actually makes the most sense for this to be in `jax.extend.ffi`, and then we can focus specifically on supporting the new XLA FFI interface, rather than general custom calls. With this in mind, I've renamed the function to `jax.extend.ffi.ffi_lowering`.
2. I updated the docstring to omit any discussion of primitives (or binding) for now. As we settle on a public API for primitives, this could be updated.

There's one inline question about supporting dynamic shapes out of the box, but I'm not very familiar with that use case. Otherwise, I think this is good to go!

superbobry commented:
Looks great, ship it!

ezhulenev commented on jax/_src/extend/ffi.py:
```diff
@@ -59,3 +67,77 @@ def include_dir() -> str:
+
+  Args:
+    call_target_name: The name of the custom call target.
+    operand_layouts: A sequence of layouts (dimension orders) for each operand.
```
FWIW I don't think this (layouts) is fully supported today, it's easy to add strides to capture buffer layout, but no one asked for it so far and we don't have any examples

dfm replied:
@ezhulenev — Oh interesting - good to know! So, if I understand correctly, the custom call always receive buffers in row-major(?) order regardless of the `layout` values passed to mlir?

rocholojavinar commented:
![](https://img.youtube.com/vi/VSxuNRF8RHY/maxresdefault.jpg)
![](https://img.youtube.com/vi/zHY8buL8RdE/maxresdefault.jpg)
![](https://img.youtube.com/vi/OFgcVINZnL4/maxresdefault.jpg)
